### PR TITLE
feat: refactor Calendar component to match Fiori 3 specs

### LINKED ISSUE #555: Calendar lighten/darken parameters
The Calendar component specifications use darken and lighten functions for the background colors of the calendar items. We need the computed values exposed as theme parameters in sap-theming.

### PR DESCRIPTION
**Pull request by JKMarkowski**
## Related Issue
Closes https://github.com/SAP/fundamental-styles/issues/504
Closes https://github.com/SAP/fundamental-styles/issues/525
Closes #555 (missing values have been delivered)

## Description
This PR introduces new version of Calendar module matching Fiori3 specification.

## Screenshots
Standard calendar:

image

image

image

Calendar compact mode:

image

Calendar on mobile device in landscape orientation:

image

Calendar on mobile device portrait orientation:

image

Calendar special days markers:

image

### CODE REVIEW

File: src/calendar.scss [hunk 1]
```diff
@@ -19,112 +19,152 @@ $block: #{$fd-namespace}-calendar;
 
 // VARS
 // making these more readily available
-$fd-calendar-item-color: fd-color-state("selected", "action") !default;
-$fd-calendar-content-background-color: fd-color("background", 2) !default;
-$fd-calendar-item-color--other-month: fd-color("text") !default;
 $fd-calendar-item-color--disabled: fd-color-state("disabled", "action") !default;
```
Comment (InnaAtanasova) [inline src/calendar.scss]: should be coming from sap-theming

File: src/calendar.scss [hunk 2]
```diff
@@ -19,112 +19,152 @@ $block: #{$fd-namespace}-calendar;
...
+  min-width: $fd-calendar-item-compact-width;
+}
 
-  opacity: 0.4;
+%fd-calendar-item-other-month {
+  background-color: transparent;
 }
 
 %fd-calendar-item-disabled {
   @include fd-var-color("color", $fd-calendar-item-color--disabled, --fd-color-action-disabled);
```
Comment (InnaAtanasova) [inline src/calendar.scss]: should be coming from sap-theming

**Comment by InnaAtanasova** [comment]
Created an issue for the darken/lighten values [here](https://github.com/SAP/fundamental-styles/issues/555)

**Comment by bcullman** [comment]
should there be before shots on the description here?

**Comment by salarenko** [comment]
Hey 
@InnaAtanasova I've updated calendar to match Fiori3 specification. Codes have changed so I'd like to ask you to once again check this PR. Unfortunately someone recently updated documentation for calendar (mobile view mostly) so I recommend to check this as well.

- I've introduced special days
- I've introduced mobile layout, but without outer container (Dialog)
- I've created examples of year ranges [2018 - 20218] in tests, I don't think showing these in documentation is necessary, but I can post them there as well.

@bcullman I don't think we need Before / After when we introduce new feature.

**Comment by droshev** [comment]
@salarenko we agreed to make the month and year view to be with the same height and put more years

File: src/calendar.scss [hunk 3]
```diff
@@ -3,178 +3,254 @@
...
+$fd-calendar-item-focused-outline-style: var(--sapContent_FocusStyle) !default;
+
+$fd-calendar-item-weekend-background: #dedede !default; // darken(--sapList_Background, 13)
+$fd-calendar-item-weekend-background-hover: #cccccc !default; // darken(--sapList_Background, 20)
+
+$fd-calendar-item-range-background: var(--sapList_SelectionBackgroundColor) !default;
+$fd-calendar-item-range-background-hover: var(--sapList_Hover_SelectionBackground) !default;
+$fd-calendar-item-range-border: solid 0.0625rem var(--sapList_SelectionBorderColor) !default;
+
+$fd-calendar-item-current-border: solid medium var(--sapAccentColor4) !default;
```
Comment (JKMarkowski) [inline src/calendar.scss]: Unfortunately meduim doesn't match the specs, we go back to `0.125rem` 
example:
image
Review:
image

File: src/calendar.scss [hunk 4]
```diff
@@ -3,178 +3,254 @@
...
+
+$fd-calendar-item-disabled-color: fd-color-state("disabled", "action") !default;
```
Comment (JKMarkowski) [inline src/calendar.scss]: Let's remove it, it's not themable

File: docs/pages/components/calendar.md [hunk 5]
```diff
@@ -18,747 +18,720 @@ Classes available for `.fd-calendar__item`. All are available to dates, months,
...
+    </div>
+</div>
+{% endcapture %}
+{% include display-component.html component=default-calendar-compact %}
+
+## Example calendar in mobile mode lanscape orientation
+Calendar should be opened inside dialog component.
+
+{% capture default-calendar-mobile %}
+<div class="fd-calendar fd-calendar--mobile" style="position: static;">
```
Comment (JKMarkowski) [inline docs/pages/components/calendar.md]: Can you add some wrapper with inline styling `max-width: 400px` there?

**Comment by JKMarkowski** [comment]
image

Buttons there should be `2rem` (32px) height, right now it's `36px`

**Comment by JKMarkowski** [comment]
image

Days should have `2.875rem (46px)` right now there is `44px`

**Comment by JKMarkowski** [comment]
image
Month/Year/increment/decrement buttons should be `2.75rem (44px)`, right now it has `36px`

**Comment by droshev** [comment]
@salarenko we need to document all the classes that were removed and/or what are their successors

**Comment by jbadan** [comment]
@jacobdevera Can you confirm that the html changes are consistent with your a11y audit?

File: docs/pages/components/calendar.md [hunk 6]
```diff
@@ -18,747 +18,784 @@ Classes available for `.fd-calendar__item`. All are available to dates, months,
...
+                    <span class="fd-calendar__text" role="button">F</span>
+                </th>
+                <th class="fd-calendar__item fd-calendar__item--side-helper">
+                    <span class="fd-calendar__text" role="button">S</span>
+                </th>
+            </tr>
+            </thead>
+            <tbody class="fd-calendar__group">
+            <tr class="fd-calendar__row">
+                <td class="fd-calendar__item fd-calendar__item--side-helper" role="gridcell">
```
Comment (jacobdevera) [inline docs/pages/components/calendar.md]: The `gridcell` role should be applied to the focusable elements in the grid (the dates). It probably doesn't hurt to leave it on the `--side-helper`s.

The rest of the markup appears consistent with the previous a11y changes.

**Comment by stefanoScalzo** [comment]
Screen Shot 2020-02-12 at 7 32 39 PM
focus looks weird on corners

**Comment by stefanoScalzo** [comment]
Should the dates be tabable for accessibility?

**Comment by stefanoScalzo** [comment]
To avoid communication removes the special days in the range selector and where they're not used

**Comment by salarenko** [comment]
@JKMarkowski I believe your issues with button sizing have been addressed.
@droshev Component bleeding has been fixed, I also updated calendar years example.
@stefanoScalzo Once again I have o ask you not to review PRs marked as [WIP].

- I cannot reproduce behavior presented on your screenshot, could you check if this still can be reproduced?
- I think grid elements navigation should be implemented on component consuming Fundamental Styles side.
> To avoid communication removes the special days in the range selector and where they're not used
- What do you mean?

**Comment by JKMarkowski** [comment]
Mobile Mode Landscape looks different than in specs. i would recommend setting max-height to same size as on portrait for `fd-calendar__navigation` class.
image
image

**Comment by salarenko** [comment]
I've had a talk with designers responsible for Calendar and navigation buttons shouldn't be modified by any calendar styles. Navigation should be composed out of default buttons components in `cosy` or `compact` mode.
Right now buttons in `cosy` mode have `height: 36px` and in `compact` mode `height: 28px` thats different as specified in docs (`44px` and `32px`) but this behavior should be eventually changed on Button component side.

File: docs/pages/components/calendar.md [hunk 7]
```diff
@@ -10,755 +10,1217 @@ summary:
...
-			</table>
-		</div>
-	</div>
+    <header class="fd-calendar__header">
+        <div class="fd-calendar__navigation">
+			<div class="fd-calendar__action fd-calendar__action--arrow-left">
+    			<button type="button" class="fd-button fd-button--light sap-icon--slim-arrow-left" aria-label="Previous"></button>
+    		</div>
+    		<div class="fd-calendar__action">
+    			<button type="button" class="fd-button fd-button--light" aria-selected="false" aria-expanded="false" aria-controls="fm3cr569">January</button>
```
Comment (JKMarkowski) [inline docs/pages/components/calendar.md]: Let's remove useless `aria` labels,  clicking on it causes 
image
error.

Comment (jacobdevera) [reply docs/pages/components/calendar.md]: The `aria-controls` seems to be the problematic property here. It doesn't appear to correctly link to any valid id on the page, either.

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-02-27 at 4 17 35 PM

The "inner" buttons have margin-left and margin-right 0.25rem and combined they get total gutter 0.25rem. But the "outer" buttons (the arrows) don't have margins which leaves the gutter to 0.25rem

**Comment by InnaAtanasova** [comment]
The cursor shouldn't be a pointer

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-02-27 at 4 31 36 PM

I think it's missing padding on the top. The buttons are touching the borders of the container. 
Just a question: should we have a footer with ok button in landscape mode?

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-02-27 at 4 34 36 PM
Missing space between `Date|Time` buttons and the header

**Comment by InnaAtanasova** [comment]
A few small things from me but overall impressive work!!!! Looks great!

**Comment by InnaAtanasova** [comment]
Please also check the Playground and the example with Mobile devices. Seems a bit off
Screen Shot 2020-02-27 at 4 56 19 PM

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-02-27 at 5 00 24 PM
I think the Header is not RTL supported

**Comment by salarenko** [comment]
@JKMarkowski @rengare @InnaAtanasova I've fixed reported issues. To handle close button inside Dialog header and Calendar navigation section I've created new modifier:
-   `.fd-calendar__close-button` - should be used inside Dialog header
-   `.fd-calendar__close-button--navigation` - should be used inside Calendar navigation

@InnaAtanasova 
- There is no vertical space between [Date | Time] and Navigation specified in the docs. I think they wanted to create it using additional space created around the button with extended touchable area.
- Docs say there should be no padding at the top of the container if calendar is displayed in mobile landscape orientation.
image
  Footer is optional for both portrait and landscape orientation.

**Comment by mikerodonnell89** [comment]
On this screenshot taken from this PR, 18 is the current day:
Screen Shot 2020-03-02 at 10 22 03 AM

However I believe the current day needs a box around it, like the 5th from this screenshot:
Screen Shot 2020-03-02 at 10 21 56 AM

**Comment by salarenko** [comment]
> On this screenshot taken from this PR, 18 is the current day:
> Screen Shot 2020-03-02 at 10 22 03 AM
> 
> However I believe the current day needs a box around it, like the 5th from this screenshot:
> Screen Shot 2020-03-02 at 10 21 56 AM

Wow, from where comes the first screenshot? It looks super weird to me.

**Comment by estelaxu3** [comment]
Just one small thing, I think there is missing a pixel in section "Example calendar in mobile mode portrait orientation", between the grey box of number 3 and 10.

Screen Shot 2020-03-02 at 2 44 52 PM

The rest looks good to me.

**Comment by estelaxu3** [comment]
After checking in with Inna, here are some discussion points we'd like to bring up:

1. The spelling in month in Fiori3 specs is in full spelling, but it is abbreviation in Styles. Do we need to align it?

Fiori3 vs. Styles:
Screen Shot 2020-03-02 at 4 15 31 PM
 
Screen Shot 2020-03-02 at 4 11 15 PM

2. There are 5 blocks of years in one line in Fiori3 but only 4 in Styles.
Fiori3 vs. Styles:
Screen Shot 2020-03-02 at 4 15 19 PM
Screen Shot 2020-03-02 at 4 15 25 PM

**Comment by salarenko** [comment]
@estelaxu3 You are right. Calendar examples have been updated to match docs examples. I've also fixed the spacing between calendar grid elements.

**Pull request merged**